### ISSUE TITLE
Bug resizing (resampling) transparent 8-bit PNG 

### ISSUE THREAD

**Issue by RoBorg** [libgd/libgd#201]
There seems to be a bug in the latest version (2.1.1) with resampling an image that was loaded from a transparent 8-bit PNG. Previously the script below worked fine, but now it ends up putting a coloured fringe where the transparent areas meet:
![zoomed-screenshot](https://cloud.githubusercontent.com/assets/702158/12850751/3b7b3d30-cc1e-11e5-93e2-8b96d8285d38.PNG)

The scripts uses "blank" image that are filled with pink at 100% transparent, so the pink colour should never show up. It seems like when it's resampled, it's doing something incorrectly which results in the pink being visible.

Test image:
![test](https://cloud.githubusercontent.com/assets/702158/12850754/3fda93a8-cc1e-11e5-8a57-324a2ae00f4b.png)

Test script:
```
<?php

$srcWidth = 80;
$srcHeight = 42;
$destWidth = 200;
$destHeight = 200;

// Create an image and fill it with true transparency
function blankImage($width, $height)
{
	$img = imagecreatetruecolor($width, $height);
	
	imagealphablending($img , false);
	$c = imagecolorallocatealpha($img, 255, 0, 255, 127);
	imagefilledrectangle($img, 0, 0, $width - 1, $height - 1, $c);

	return $img;
}

// Merge layers on top of each other
function flatten($layers, $width, $height)
{
	$img = blankImage($width, $height);
	imagealphablending($img, true);
	
	foreach ($layers as $l)
		imagecopy($img, $l, 0, 0, 0, 0, $width, $height);
	
	imagesavealpha($img, true);
	
	return $img;
}

$background = blankImage($destWidth, $destHeight);

$logoSource = imagecreatefrompng(__DIR__ . '/test.png');
//imagepalettetotruecolor($logoSource);

$logo = blankImage($srcWidth, $srcHeight);
imagealphablending($logo, false);
imagecopy($logo, $logoSource, 0, 0, 0, 0, $srcWidth, $srcHeight);

$scaledLogo = blankImage($destWidth, $destHeight);
imagealphablending($scaledLogo, false);
imagecopyresampled($scaledLogo, $logo, 0, 0, 0, 0, 200, 105, $srcWidth, $srcHeight);

$img = flatten(array($background, $scaledLogo), $destWidth, $destHeight);

header('Content-type: image/png');
imagepng($img);
```

**Comment by pierrejoye** [comment]
Does it happen with imagescale too?

**Comment by RoBorg** [comment]
Yes, in fact it seems to be considerably worse
![imagescale](https://cloud.githubusercontent.com/assets/702158/15747009/b2a740e8-28d0-11e6-92f5-5501926bd20f.png)

[test.txt](https://github.com/libgd/libgd/files/295726/test.txt)

**Comment by pierrejoye** [comment]
What happened is the alpha channel is also taken into account in the extrapolation. That means the pink become visible as the transparency is not 100% anymore.

There are cases where it makes results way better and such behaviors are expected. Clearly not in this case.

I am not sure how to provide an automatic detection in this case as the result is rather undesired.

@nathanaeljones ideas how to deal with such cases? It is easy to ignore the fully transparent pixels but can lead to wrong results in many cases.

**Comment by pierrejoye** [comment]
Example using the fuel logo only (which contains the transparency):

$logoSource = imagecreatefrompng(__DIR__ . '/gd201.png');
$im = imagescale($logoSource, $destWidth, $destHeight);
imagepng($im, __DIR__.'/res.png');

shows the correct behavior for example. The alpha is respected and scaled accordingly.

![res](https://cloud.githubusercontent.com/assets/282408/15766063/538a65dc-2965-11e6-8b82-7c2659769b26.png)

**Comment by pierrejoye** [comment]
Also using master (php) and bundled GD, the behavior is fixed for your case as well. I was first trying to reproduce the actual problem using a simple case.

See the image below which is generated using your script:

![res](https://cloud.githubusercontent.com/assets/282408/15766196/216b1186-2966-11e6-89a1-807f9c417924.png)

I suppose you use 5.x and external GD 2.1.1? If you could try with 2.2.1 or I can try later today as well.

**Comment by RoBorg** [comment]
I'm on holiday at the moment so I can't test anything with different versions for a while :frowning:

**Comment by cmb69** [comment]
That appears to be a libgd issue (reproduced with current master), that has been fixed in PHP's bundled libgd long ago (tested with 5.4.19). I'll investigate.

**Comment by cmb69** [comment]
I've found the respective [commit for PHP's bundled libgd](https://github.com/php/php-src/commit/8f6d5dc775cbc968199163465bd0111236d95a96) (it's hard to read, but look for `alpha_factor`). Will look for respective changes in imagescale() ASAP.

**Comment by pierrejoye** [comment]
@cmb69 By the way. ImageCopyResized and ImageCopyResampled has nothing to do with ImageScale. I do not think this commit is related to what is described here.


The problem described here is that the channels (r,g,b) should not be taken into account during scaling if transparency is 100%. The possible issue (to be discussed if it is an issue) is that the target pixel can be  the result of the medium values of many pixels, at the end, even if some pixels have 100% transparencey, the resulting pixel may have a different transparency value (f.e., 4 pixels used for the target, 2 has 0%, 2 has 100%).

Also see my comment https://github.com/libgd/libgd/issues/201#issuecomment-223469603 as in, I cannot reproduce it :)

Do you have a C code to reproduce it please?

**Comment by cmb69** [comment]
> Do you have a C code to reproduce it please?

I've submitted the respective PR #263 with the failing test case (transcribed from the OP's PHP code). The contained bug00201_exp.png has been generated with the bundled libgd of PHP 5.4.19 (newer bundled GDs would also have worked).

**Comment by pierrejoye** [comment]
Thanks :)

Do you see the bug too with gdImageScale? I did not. The reporter mentioned
it was failing in an even more bad way but for me it was working as
expected.

**Comment by cmb69** [comment]
> Do you see the bug too with gdImageScale?

Yes, I can reproduce that (again only with external libgd, not PHP's bundled one). I've attached a respective patch as comment to PR #263.

**Comment by cmb69** [comment]
> Yes, I can reproduce that (again only with external libgd, not PHP's bundled one).

No. That also happens with PHP's bundled libgd (tested with PHP 7.0.8).

AIUI, libgd stores unassociated RGBA internally, and the resampling uses these values directly. Shouldn't associated RGBA (i.e. the color values multiplied with the opacity percentage) be used during the resampling? It seems to me that this (or something like this) is done by PHP's bundled libgd in imagecopyresampled().

**Comment by pierrejoye** [comment]
On Jul 19, 2016 12:20 AM, "Christoph M. Becker" <notifications@github.com>
wrote:
>>
>> Yes, I can reproduce that (again only with external libgd, not PHP's
bundled one).
>
> No. That also happens with PHP's bundled libgd (tested with PHP 7.0.8).
>
> AIUI, libgd stores unassociated RGBA internally, and the resampling uses
these values directly. Shouldn't associated RGBA (i.e. the color values
multiplied with the opacity percentage) be used during the resampling? It
seems to me that this (or something like this) is done by PHP's bundled
libgd in imagecopyresampled().

That causes issues when you want to scale image with alpha!=100.

If you have something like

80 76
58 100

And you scale down, I expect a medium value for alpha here.

However the correct fix in the bundled gd should have been different. But I
agree that the way gd handles background color and alpha can be confusing.
They are different things.

Only for the record, do you consider the behavior of imagescale correct or
is it also unexpected?

**Comment by cmb69** [comment]
> That causes issues when you want to scale image with alpha!=100.

For instance, we have two pixels with RGBA(0, 0, 0, 1) (solid black) and RGBA(1, 0, 0, 0.1) (mostly transparent red). We calculate the associative alpha values (i.e. multiply the colors by alpha) as RGBA(0, 0, 0, 1) and RGBA(0.1, 0, 0, 0.1), respectively. Then we resample (bilinear) to RGBA(0.05, 0, 0, 0.55). Then we unassociate (i.e. divide the colors by alpha) to RGBA(0.09, 0, 0, 0.55) – a slighty reddish, very dark, half-transparent pixel. Looks good to me.

> Only for the record, do you consider the behavior of imagescale correct or
is it also unexpected?

IMHO, that's a bug. At least it is unusable for this case (and most likely for many others as well).

**Comment by pierrejoye** [comment]
>> Only for the record, do you consider the behavior of imagescale correct or is it also unexpected?
>  IMHO, that's a bug. At least it is unusable for this case (and most likely for many others as well).

My simple test case works. I do not see what is wrong with the results (see my reply in this bug https://github.com/libgd/libgd/issues/201#issuecomment-223469603 and the previous one on the same day).

You consider these two as wrong, do I understand you correctly? 

If yes, then I failed to see how these examples are wrong (they render wrong here but try it  and you will see correct interpolated transparency and colors).

**Comment by cmb69** [comment]
> If yes, then I failed to see how these examples are wrong (they render wrong here but try it and you will see correct interpolated transparency and colors).

I'm afraid I don't understand. The test case in PR #263 fails on my local machine and also [on Travis](https://travis-ci.org/libgd/libgd/jobs/145585842#L1442). The result I'm getting:

![bug00201 c_68_out](https://cloud.githubusercontent.com/assets/2306138/16925720/d2420c02-4d25-11e6-9bb7-c3021777b1f4.png)

gdImageScale() is even worse:

![bug00201 c_71_out](https://cloud.githubusercontent.com/assets/2306138/16925737/e19d8ef6-4d25-11e6-8cf8-47a86b192342.png)

The pink halo around the glyphs appears to be the result of resampling RGBA(0, 0, 0, 0) and RGBA(1, 0, 1, 1) to RGBA(0.5, 0, 0.5, 0.5), what is wrong IMHO.

The result I would expect (and that I get with PHP's bundled libgd using imagecopyresampled(), but not imagescale(), and never with "our" libgd):

![bug00201_exp](https://cloud.githubusercontent.com/assets/2306138/16925876/9bd4f804-4d26-11e6-9306-3590033346f1.png)

**Comment by pierrejoye** [comment]
Hm. Weird.

However I stand wrong and you are right about the premultiplied.

Here is a lengthy explanation

http://blog.wolfire.com/2009/01/scaling-images-with-alpha/

I will look into imagescale and see how to fix it. But that will be post
2.2.3 due on Tomorrow or Wednesday.

**Comment by cmb69** [comment]
> I will look into imagescale and see how to fix it.

Thanks. Other code might be affected as well (imageconvolution() comes to mind). Anyway, that's probably lot of work – if I can help, just say so. :-)

BTW: *storing* premultiplied RGBA *might* also be an option, to improve the performance of multiple scaling/resampling etc. operations. A flag could be added to gdImage that tells whether tpixels are associative or unassociative colors. But that could be postponed even further, as it would only be an optimization (if at all; I guess we'd need some benchmarks).

**Comment by pierrejoye** [comment]
And here the patch based on your finding :)

Please wait before applying, we need to test how it behaves with random alpha values, to be 100% sure.

[imagecopyresampled_premultiplied.txt](https://github.com/libgd/libgd/files/370642/imagecopyresampled_premultiplied.txt)

As of using premultiplied surface, it is planed for post 2.2, as part of true 32bit surface (with 255 alpha, not 127, or not 32-1bit for the 4 channels). I would for now only focus on scale and eventually rotate (to check if also affected but I suppose it is).

**Comment by pierrejoye** [comment]
@cmb69  I think we can have that fix in 2.2.3. Do you have time to merge yourself? I like to relase 2.2.3 today/tomorrow

Better to open a new bug for the other cases.

**Comment by cmb69** [comment]
> I think we can have that fix in 2.2.3. Do you have time to merge yourself?

Will have a look at it later today.

**Comment by pierrejoye** [comment]
WIll wait on that one until midnight (in 7h), if not we can do it for 2.2.4. :)

**Comment by cmb69** [comment]
@pierrejoye I would suggest to only address gdImageCopyResampled() for 2.2.3, namely to make it match PHP's bundled libgd (less confusion for developers and bug handlers). Your patch does this, sans some minor differences which may result in different rounding; your patch uses floats, PHP's bundled libgd uses double; bundled libgd uses the floor_cast macro, albeit, apparently, inconsistently; sx1, sx2, sy1 and sy2 are calculated in different places (bundled libgd is therefore a tad faster). I don't mind these differences (at least for 2.2.3), but there is one major difference, namely an [additional modification of alpha](https://github.com/php/php-src/blob/PHP-7.0.8/ext/gd/libgd/gd.c#L2547). I'm not sure, what's the correct solution. I'll add some test cases, which hopefully will point to the better algo.

I would leave the other cases (scale, rotate etc.) for 2.2.4, so we have enough time to come up with a good solution and also adapt PHP's bundled libgd accordingly.

**Comment by pierrejoye** [comment]
Yes, you are right.

Syncing first with 2.2.3 then move to a overall better solution for quality
and consistency.

**Comment by cmb69** [comment]
I have noticed that the patch has already been merged, so I have now added a few test cases for gdImageCopyResampled(), most notably [exact_alpha](https://github.com/libgd/libgd/blob/master/tests/gdimagecopyresampled/exact_alpha.c). This test succeeds because I have adapted the expected values to the actual values. I would, however, expect different values, namely red==50, green== 50, blue==100 (alpha==32 is already fine). Anyhow, the actual values are not so bad (the old algo would probably have reported 100, 50, 125), so we might leave it as it is for now.
